### PR TITLE
Push API updates to simplify code, better error handling, add support for albums

### PR DESCRIPTION
**Pull request by xavierbeynon**
* Updated the api call to return more JSON info (as opposed to page scraping)
* Caught additional inconsistent error messages
* Added cases to fill in missing metadata with mediocre values, as opposed to failing all together
* Added audiomack album/playlist class to support album downloads

### CODE REVIEW

File: youtube_dl/extractor/audiomack.py [hunk 1]
```diff
@@ -9,20 +9,21 @@
 
 
 class AudiomackIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?audiomack\.com/song/(?P<id>[\w/-]+)'
+    _VALID_URL = r'https?://(?:www\.)?audiomack\.com/(song)/(?P<id>[\w/-]+)'
```
Comment (phihag) [inline youtube_dl/extractor/audiomack.py]: Why is this change correct? This looks really strange and just adds a useless group, doesn't it?

File: youtube_dl/extractor/audiomack.py [hunk 2]
```diff
@@ -9,20 +9,21 @@
...
     IE_NAME = 'audiomack'
     _TESTS = [
-        # hosted on audiomack
+        # audiomack
```
Comment (phihag) [inline youtube_dl/extractor/audiomack.py]: I very much prefer the older comment, which was much clearer.

File: youtube_dl/extractor/audiomack.py [hunk 3]
```diff
@@ -9,20 +9,21 @@
...
         {
             'url': 'http://www.audiomack.com/song/roosh-williams/extraordinary',
             'info_dict':
             {
-                'id': 'roosh-williams/extraordinary',
-                'ext': 'mp3',
-                'title': 'Roosh Williams - Extraordinary'
+                'id': '310086',
+                "ext": "mp3",
```
Comment (phihag) [inline youtube_dl/extractor/audiomack.py]: Why have you changed the quotes here? It's not that important, but we strive to use `'` when possible, and have when possible a consistent quote character in a file.

File: youtube_dl/extractor/audiomack.py [hunk 4]
```diff
@@ -37,33 +38,88 @@ class AudiomackIE(InfoExtractor):
         },
     ]
 
+    @staticmethod
+    def create_song_dictionary(api_response, album_url_tag, track_no=0):
+        # All keys are the same in audiomack api and InfoExtractor format
```
Comment (phihag) [inline youtube_dl/extractor/audiomack.py]: This seems to be a really complicated way of writing
```
{
    'title': api_response.get('title', album_url_tag),
    'uploader': api_response.get('artist'),
    'url': api_response['url'],
}
```
Notably, `id` looks really strange - shouldn't that include the `video_id`?

File: youtube_dl/extractor/audiomack.py (hunk 1 above)
Comment (xavierbeynon) [inline youtube_dl/extractor/audiomack.py]: Sorry, that was a typo.  Originally I was detecting whether it was song or album, before I broke album off into its own extractor.

File: youtube_dl/extractor/audiomack.py (hunk 2 above)
Comment (xavierbeynon) [inline youtube_dl/extractor/audiomack.py]: Agreed.

File: youtube_dl/extractor/audiomack.py (hunk 3 above)
Comment (xavierbeynon) [inline youtube_dl/extractor/audiomack.py]: I made them consistent.

File: youtube_dl/extractor/audiomack.py (hunk 4 above)
Comment (xavierbeynon) [inline youtube_dl/extractor/audiomack.py]: I didn't change this.  Yes, the dictionary comprehension could be unraveled.  But I think the one liner followed by the if statements makes clear what is expected, followed by what must be done with the unexpected.  Both are clear and equivalent; mine is just a bit more fun.

I don't know what you mean by the id's.  I understand that an id field must be present in the dictionary - the API response from AM conveniently contains a corresponding field that I copy over.

**Comment by phihag** [comment]
Thanks! Merged.

**Pull request merged**